grakn-graph/src/main/java/ai/grakn/graph/internal/RelationTypeImpl.java

```diff
@@ -80,6 +76,13 @@ public Relation addRelation() {
     public RelationType hasRole(RoleType roleType) {
         checkTypeMutation();
         putEdge(roleType, Schema.EdgeLabel.HAS_ROLE);
+
+        //Cache the Role internally
+        cachedHasRoles.get().add(roleType);
```

Comment:
In some cases we're using `isPresent` before doing this, though. Is there a rule for when we do that check and when we don't?